### PR TITLE
Fixes duplicate version number + others suggestions.

### PR DESCRIPTION
**Pull request by jtkech**
I tried to use the simplest way.

1. In `DefaultContentManagerSession` which is a scoped service, items are cached by various states, `Id`, `ContentItemId` & `Number`, `Published`. In a previous PR i was thinking to remove an item from the cache when the related state is updated. But finally, for simplicity we can assume that the methods which modify these states are intended to be called at the end of a request. **So, that's ok for me**.

2. In `BuildNewVersionAsync()`, even it is only called on a `Latest and Published` item (which is session saved before), we still check if it is the `Latest` and if not we are doing a `query`. So, assuming that this could be possible in the future, because we modify the queried item, i added a `_session.Save()`.

3. Then, to fix the duplicate version number which happens after having discarding a draft, so the `Published` item becomes the `Latest` but not with the highest `Number`, here the suggestion is to simply also assign to this `Published` item the `Number` of the draft we are discarding + 1.

### CODE REVIEW

**Comment by sebastienros** [comment]
Do we actually need this `Number` property? I think that as we introduced `ContentItemVersionId` it's not necessary anymore. Can you check?

**Comment by jtkech** [comment]
Good question, i thought about this but i didn't want to "complicate" things.

- We use it in `DefaultContentManagerSession` to populate the `_itemByContentItemId` where we use a tuple of `ContentItemId` & `Number`. But currently we never use `RecallContentItemId()`. So, maybe we could remove this cache.

- We check if `Number` > 0 in the `OwnerEditorDriver` to see if it's not a new item. A new item has its `Number` equal to 0 but its `ContentItemVersionId` has been generated. Oh, but here we could use the new `context.IsNew` that you created and we used for fields.

That said, maybe useful to keep an incremental version `Number`. Maybe a scenario where we would want to retrieve a previously created item, e.g in my versioning PR. Otherwise i don't see any other usage, so, if it can make things simpler i can do it, it's as you want.

**Comment by jtkech** [comment]
Removing the `Number` prop, just saw that it is used at the beginning of the `DefaultContentManager` `Create()` method to see if it is a new item, then if so we init `Latest = true` and `Published = true`.

Hmm, maybe not necessary, i will check.

**Comment by jtkech** [comment]
Finally i like the idea of removing the `Number` property.

So i did it, can be reverted just to show it.

Only one doubt at the beginning of the content manager `Create()` method.

            if (contentItem.Number == 0)
            {
                contentItem.Number = 1;
                contentItem.Latest = true;
                contentItem.Published = true;
            }

Not sure i understand, is it possible to do a `Create()` by passing an item with `Number != 0`?

If it is only to check if the item has no related database record, i think the following is ok.

            if (contentItem.Id == 0)
            {
                contentItem.Latest = true;
                contentItem.Published = true;
            }

**Comment by sebastienros** [comment]
Don't use `Id`. Check where `Create` is called, and maybe it just need a boolean `isNew` instead, or `isLatest` and `isPublished`.

**Comment by jtkech** [comment]
Ok, i will check.

**Comment by jtkech** [comment]
Done but not sure at 100%.

Most of the time, a content manager `New()` is done before `Create()` so i pass `isNew = true`.
Then, in `ContentStep.ExecuteAsync()` and in `ContentApiController.AddContent()` i pass `isNew = false`.

I've tried a new blog setup to check the blog recipe, and created some new articles.

Note: hmm, it seems that for menus you don't use content manager `Create()`.

**Comment by sebastienros** [comment]
Looks like it's always `true` unless `VersionOption` is specific and some values tell that it's not `latest` and `published`.

**Comment by sebastienros** [comment]
So you just need `Create` with and without options, then based on the option define what `published` and `latest` are.

**Comment by jtkech** [comment]
Okay but maybe it depends on the possible scenarios.

- `Latest option` => `Latest = true`, then `Published = ?`. I would say `Published = false` but then this option becomes the same as the following draft option.

- `Draft option` => `Published = false`, then `Latest = true` otherwise it would be auto soft deleted.

- `Published option` => `Published = true`, then `Latest = ?`. I would say `Latest = true` but maybe a scenario (e.g through a recipe) where you want to create a published item which is not the latest.

So, i would suggest to always set `Latest = true` and `Published = true` at the beginning and then reset `Published` to false if the draft option is specified, and check the published option to invoke handlers.

So, we would only need to remove the `isNew` parameter and its checking. So, it would be as before but without the `Number` checking. Hmm, maybe we loose a scenario but it's ok for me if you agree.

I will update it as suggested here.

File: src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs [hunk 1]
```diff
@@ -320,18 +319,13 @@ public void Create(ContentItem contentItem)
 
         public void Create(ContentItem contentItem, VersionOptions options)
         {
-            if (contentItem.Number == 0)
+            if (String.IsNullOrWhiteSpace(contentItem.ContentItemVersionId))
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs]: `IsNullOrEmpty`

**Pull request merged**